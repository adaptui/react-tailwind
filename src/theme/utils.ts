import { defaults, isString, isUndefined, mergeWith, tcm } from "../utils";

import {
  DefaultTheme,
  ExtendableDefaultTheme,
  PartialDefaultTheme,
} from "./types";

export const extendTheme = <T = ExtendableDefaultTheme>(theme: T): T => {
  return theme;
};

// Based on https://github.com/tailwindlabs/tailwindcss/blob/6ce8864f47061035cd79b2d563eea72467dcb3b7/src/util/resolveConfig.js
function collectExtends(items: ExtendableDefaultTheme[]) {
  return items.reduce((merged, { extend }) => {
    return mergeWith(merged, extend, (mergedValue, extendValue) => {
      if (isUndefined(mergedValue)) {
        return [extendValue];
      }

      if (Array.isArray(mergedValue)) {
        return [extendValue, ...mergedValue];
      }

      return [extendValue, mergedValue];
    });
  }, {});
}

// Based on https://github.com/tailwindlabs/tailwindcss/blob/6ce8864f47061035cd79b2d563eea72467dcb3b7/src/util/resolveConfig.js
export function mergeThemes(themes: PartialDefaultTheme[]) {
  return {
    ...themes.reduce<DefaultTheme>(
      (merged, theme) => defaults(merged, theme),
      {} as DefaultTheme,
    ),

    // In order to resolve n config objects, we combine all of their `extend` properties
    // into arrays instead of objects so they aren't overridden.
    extend: collectExtends(themes),
  };
}

// Based on https://github.com/tailwindlabs/tailwindcss/blob/6ce8864f47061035cd79b2d563eea72467dcb3b7/src/util/resolveConfig.js
export function mergeExtensions({
  extend,
  ...theme
}: DefaultTheme & { extend?: PartialDefaultTheme }) {
  return mergeWith(theme, extend, (themeValue, extendValue) => {
    return mergeWith(
      themeValue,
      ...extendValue,
      (merged: any, value: any[]) => {
        if (isString(merged) && isString(value)) {
          return tcm(merged, value);
        }

        return undefined;
      },
    );
  });
}
