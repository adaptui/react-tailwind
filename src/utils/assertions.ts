import { Dict } from "./types";

// Null / Undefined assertions
export function isUndefined(value: any): value is undefined {
  return typeof value === "undefined" || value === undefined;
}

export { isNull } from "@renderlesskit/react";

// String assertions
export function isString(value: any): value is string {
  return Object.prototype.toString.call(value) === "[object String]";
}

// Array assertions
export function isArray<T>(value: any): value is Array<T> {
  return Array.isArray(value);
}

// Object assertions
export const isObject = (value: any): value is Dict => {
  const type = typeof value;

  return value != null && type === "object" && !isArray(value);
};

// Function assertions
export function isFunction(value: any): value is Function {
  return typeof value === "function";
}

// SSR check
export function canUseDOM() {
  return !!(
    typeof window !== "undefined" &&
    window.document &&
    window.document.createElement
  );
}

export const isBrowser = canUseDOM();

export function isEmptyObject(value: any) {
  return isObject(value) && Object.keys(value).length === 0;
}
