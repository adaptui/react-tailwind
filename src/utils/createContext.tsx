import * as React from "react";
import { isUndefined } from "lodash";

export interface CreateContextOptions<ContextType> {
  /**
   * The display name of the context
   */
  name?: string;

  /**
   * Default context value.
   */
  defaultContext?: ContextType;

  /**
   * If `true`, React will throw if context is `null` or `undefined`
   * In some cases, you might want to support nested context, so you can set it to `false`
   */
  strict?: boolean;

  /**
   * Error message to throw if the context is `undefined`
   */
  errorMessage?: string;
}

/**
 * Creates a named context, provider, and hook.
 *
 * @param options create context options
 */
export function createContext<ContextType extends object | null>(
  options: CreateContextOptions<ContextType> = {},
) {
  const {
    name = "Provider",
    defaultContext,
    strict = true,
    errorMessage = "Seems you forgot to wrap `useContext` within the Context Provider",
  } = options;

  const Context = React.createContext<ContextType | undefined>(defaultContext);

  function Provider(props: ContextType & { children: React.ReactNode }) {
    const { children, ...context } = props;

    // Idea to memoize within the createContext from [radix ui](https://github.com/radix-ui/primitives/blob/cefcb6d9281cf203b0ab54b49f85725dc203df85/packages/react/context/src/createContext.tsx#L13)
    // Only re-memoize when prop values change
    const value = React.useMemo(
      () => context,
      // eslint-disable-next-line react-hooks/exhaustive-deps
      Object.values(context),
    ) as ContextType;

    return <Context.Provider value={value}>{children}</Context.Provider>;
  }

  Provider.displayName = name;

  function useContext() {
    const context = React.useContext(Context);

    if (context) return context;

    if (!isUndefined(defaultContext)) return defaultContext;

    if (strict) {
      throw new Error(errorMessage);
    }

    return undefined;
  }

  return [Provider, useContext] as const;
}
