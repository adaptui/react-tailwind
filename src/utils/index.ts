export * from "./assertions";
export * from "./createContext";
export * from "./debounce";
export * from "./global-types";
export * from "./object";
export * from "./react";
export * from "./tailwindMerge";
export * from "./types";
