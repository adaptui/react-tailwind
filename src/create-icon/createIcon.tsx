// Credits to https://github.com/chakra-ui/chakra-ui/tree/main/packages/icon
import * as React from "react";

import { forwardRefWithAs, PropsWithAs } from "../utils";

import { Icon, IconProps } from "./Icon";

export interface CreateIconOptions {
  /**
   * The icon `svg` viewBox
   *
   * @default "0 0 24 24"
   */
  viewBox?: string;
  /**
   * The `svg` path or group element
   *
   * @type React.ReactNode | React.ReactNode[]
   */
  path?: React.ReactNode | React.ReactNode[];
  /**
   * If the has a single path, simply copy the path's `d` attribute
   */
  d?: string;
  /**
   * The display name useful in the dev tools
   */
  displayName?: string;
  /**
   * Default props automatically passed to the component; overwriteable
   */
  defaultProps?: PropsWithAs<IconProps, "svg">;
}

export function createIcon(options: CreateIconOptions) {
  const {
    viewBox = "0 0 24 24",
    d: pathDefinition,
    path,
    displayName,
    defaultProps = {},
  } = options;

  const IconComponent = forwardRefWithAs<IconProps, HTMLOrSVGElement, "svg">(
    (props, ref) => {
      return (
        // @ts-ignore
        <Icon ref={ref} viewBox={viewBox} {...defaultProps} {...props}>
          {path ?? <path fill="currentColor" d={pathDefinition} />}
        </Icon>
      );
    },
  );

  IconComponent.displayName = displayName;

  return IconComponent;
}
