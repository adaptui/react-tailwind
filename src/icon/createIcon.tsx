import * as React from "react";

import { Icon, IconProps } from "./Icon";
import { forwardRefWithAs, PropsWithAs } from "../utils/types";

export interface CreateIconOptions {
  /**
   * The icon `svg` viewBox
   * @default "0 0 24 24"
   */
  viewBox?: string;
  /**
   * The `svg` path or group element
   * @type React.ReactElement | React.ReactElement[]
   */
  path?: React.ReactElement | React.ReactElement[];
  /**
   * If the has a single path, simply copy the path's `d` attribute
   */
  d?: string;
  /**
   * The display name useful in the dev tools
   */
  displayName?: string;
  /**
   * Default props automatically passed to the component; overwriteable
   */
  defaultProps?: PropsWithAs<IconProps, "svg">;
}

export function createIcon(options: CreateIconOptions) {
  const {
    viewBox = "0 0 24 24",
    d: pathDefinition,
    path,
    defaultProps = {},
  } = options;

  function IconComponent(
    props: PropsWithAs<IconProps, "svg">,
    ref: React.Ref<HTMLOrSVGElement>,
  ) {
    return (
      <Icon ref={ref} viewBox={viewBox} {...defaultProps} {...props}>
        {path ?? <path fill="currentColor" d={pathDefinition} />}
      </Icon>
    );
  }

  return forwardRefWithAs<IconProps, "svg">(IconComponent);
}
