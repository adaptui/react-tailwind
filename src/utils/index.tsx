import * as React from "react";
import { createTailwindMerge } from "tailwind-merge";

import { cx } from "../utils";

import {
  As,
  ComponentWithAs,
  Dict,
  PropsWithAs,
  RenderPropType,
} from "./types";

export interface CreateContextOptions {
  /**
   * If `true`, React will throw if context is `null` or `undefined`
   * In some cases, you might want to support nested context, so you can set it to `false`
   */
  strict?: boolean;
  /**
   * Error message to throw if the context is `undefined`
   */
  errorMessage?: string;
  /**
   * The display name of the context
   */
  name?: string;
}

export type CreateContextReturn<T> = [
  React.Provider<T>,
  () => T,
  React.Context<T>,
];

/**
 * Creates a named context, provider, and hook.
 *
 * @param options create context options
 */
export function createContext<ContextType>(options: CreateContextOptions = {}) {
  const {
    strict = true,
    errorMessage = "useContext: `context` is undefined. Seems you forgot to wrap component within the Provider",
    name,
  } = options;

  const Context = React.createContext<ContextType | undefined>(undefined);

  Context.displayName = name;

  function useContext() {
    const context = React.useContext(Context);

    if (!context && strict) {
      throw new Error(errorMessage);
    }

    return context;
  }

  return [
    Context.Provider,
    useContext,
    Context,
  ] as CreateContextReturn<ContextType>;
}

// Function assertions
export function isFunction(value: any): value is Function {
  return typeof value === "function";
}

// From Chakra Utils
export function runIfFn<T, U>(
  valueOrFn: T | ((...fnArgs: U[]) => T),
  ...args: U[]
): T {
  return isFunction(valueOrFn) ? valueOrFn(...args) : valueOrFn;
}

export function runIfFnChildren<T, U>(
  valueOrFn: T | ((...fnArgs: U[]) => T),
  ...args: U[]
): T {
  // Know bug but this is the only way to make it work
  // {({ isChecked }) => {
  //   return (
  //     <>
  //       <SwitchText data-testid="testid-mode">
  //         {isChecked ? DARK_MODE : LIGHT_MODE}
  //       </SwitchText>
  //     </>
  //   );
  // }}
  // Need a wrapper to get the children when the children is a function
  // @ts-ignore
  return isFunction(valueOrFn) ? valueOrFn(...args).props.children : valueOrFn;
}

/**
 * Gets only the valid children of a component,
 * and ignores any nullish or falsy child.
 *
 * @param children the children
 */
export function getValidChildren(children: React.ReactNode) {
  return React.Children.toArray(children).filter(child =>
    React.isValidElement(child),
  ) as React.ReactElement[];
}

// SSR check
export function canUseDOM() {
  return !!(
    typeof window !== "undefined" &&
    window.document &&
    window.document.createElement
  );
}

export const isBrowser = canUseDOM();

// Array assertions
export function isArray<T>(value: any): value is Array<T> {
  return Array.isArray(value);
}

// Object assertions
export const isObject = (value: any): value is Dict => {
  const type = typeof value;

  return value != null && type === "object" && !isArray(value);
};

export const objectKeys = <T extends Dict>(obj: T) =>
  Object.keys(obj) as unknown as (keyof T)[];

// Corresponds to 10 frames at 60 Hz.
// A few bytes payload overhead when lodash/debounce is ~3 kB and debounce ~300 B.
export const debounce = (func: Function, wait = 166) => {
  let timeout: any;
  function debounced(this: any, ...args: any) {
    // eslint-disable-next-line consistent-this
    const that = this;
    const later = () => {
      func.apply(that, args);
    };
    clearTimeout(timeout);
    timeout = setTimeout(later, wait);
  }

  debounced.clear = () => {
    clearTimeout(timeout);
  };

  return debounced;
};

// String assertions
export function isString(value: any): value is string {
  return Object.prototype.toString.call(value) === "[object String]";
}

// Merge library & user prop
export const passProps = (icon: RenderPropType, props?: Dict) => {
  return React.isValidElement(icon)
    ? React.cloneElement(icon, {
        ...props,
        ...icon.props,
        className: cx(props?.className, icon.props.className),
      })
    : runIfFn(icon, { ...props });
};

// Add a11y to the icon passed
export const withIconA11y = (icon: RenderPropType, props?: Dict) => {
  return passProps(icon, {
    role: "img",
    focusable: false,
    "aria-hidden": true,
    ...props,
  });
};

export function isUndefined(value: any): value is undefined {
  return typeof value === "undefined" || value === undefined;
}

export const tcm = createTailwindMerge(getDefaultConfig => {
  const defaultConfig = getDefaultConfig();

  return {
    ...defaultConfig,
    classGroups: {
      ...defaultConfig.classGroups,
      "font-size": [
        ...defaultConfig.classGroups["font-size"],
        // Defined custom text presets
        {
          text: ["cxs", "paragraph-cxs", "paragraph-sm", "2base"],
        },
      ],
    },
  };
});

export const getComponentProps = <T, P>(
  componentMaps: Dict<string>,
  children: T,
  props: P,
) => {
  const normalizedChildren = runIfFnChildren(children, props);
  const validChildren = getValidChildren(normalizedChildren);
  const componentProps: Dict = {};
  const finalChildren: React.ReactNode[] = [];

  validChildren.forEach(child => {
    // @ts-ignore
    if (componentMaps[child.type.displayName]) {
      componentProps[
        // @ts-ignore
        componentMaps[child.type.displayName]
      ] = child.props;
    } else {
      finalChildren.push(child);
    }
  });

  return { componentProps, finalChildren };
};

/**
 * @template Props Component Props
 * @template RefProp HTML intrinsic type, eg: HTMLDivElement
 * @template DefaultType string, eg: "div"
 * @param component
 */
export function forwardRefWithAs<
  Props,
  RefProp = any,
  DefaultType extends As = any,
>(
  component: React.ForwardRefRenderFunction<
    RefProp,
    PropsWithAs<Props, DefaultType>
  >,
) {
  return React.forwardRef(component) as unknown as ComponentWithAs<
    Props,
    DefaultType
  >;
}

export * from "./types";
export { cx } from "@renderlesskit/react";
