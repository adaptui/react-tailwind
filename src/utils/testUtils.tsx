import {
  render as RtlRender,
  RenderOptions,
  waitFor,
  RenderResult,
} from "@testing-library/react";
import * as React from "react";
import { RunOptions } from "axe-core";
import { configureAxe } from "jest-axe";

// @ts-ignore
import tailwindConfig from "../../tailwind.config";
import { RenderlesskitProvider } from "../theme";
export * from "@testing-library/react";

type Render = (
  children: React.ReactNode,
  options?: RenderOptions,
) => RenderResult;

export const render: Render = (children, options = {}) => {
  return RtlRender(
    <RenderlesskitProvider tailwindConfig={tailwindConfig}>
      {children}
    </RenderlesskitProvider>,
    options,
  );
};

export const axe = configureAxe({
  rules: {
    // disabled landmark rules when testing isolated components.
    region: { enabled: false },
  },
});

type UI = Parameters<typeof render>[0] | Element;
type TestA11YOptions = RenderOptions & {
  axeOptions?: RunOptions;
};

export const testA11y = async (
  ui: UI,
  { axeOptions, ...options }: TestA11YOptions = {},
) => {
  jest.useRealTimers();

  const container = React.isValidElement(ui)
    ? render(ui, options).container
    : ui;

  // @ts-ignore
  const results = await axe(container, axeOptions);

  expect(results).toHaveNoViolations();
};

const DELAY = 0;
const LOAD_IMAGE = "load.png";
const ERROR_IMAGE = "error.png";
const orignalImage = window.Image;

export const mockImage = (loadState: string) => {
  jest.useFakeTimers();

  (window.Image as unknown) = class MockImage {
    onload: () => void = () => {};
    onerror: () => void = () => {};
    src: string = "";
    constructor() {
      if (loadState === LOAD_IMAGE) {
        setTimeout(() => {
          this.onload();
        }, DELAY);
      }
      if (loadState === ERROR_IMAGE) {
        setTimeout(() => {
          this.onerror();
        }, DELAY);
      }
      return this;
    }
  };
};

mockImage.DELAY = DELAY;
mockImage.LOAD_IMAGE = LOAD_IMAGE;
mockImage.ERROR_IMAGE = ERROR_IMAGE;
mockImage.ERROR_IMAGE = ERROR_IMAGE;
mockImage.load = () => mockImage(LOAD_IMAGE);
mockImage.error = () => mockImage(ERROR_IMAGE);
mockImage.restoreMock = () => (window.Image = orignalImage);
mockImage.advanceTimer = async () => {
  return await waitFor(() => {
    jest.advanceTimersByTime(mockImage.DELAY);
  });
};
