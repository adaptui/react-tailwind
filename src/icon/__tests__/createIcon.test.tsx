import { createIcon } from "../createIcon";
import { render, testA11y } from "../../utils/testUtils";

const Test = createIcon({
  displayName: "CloseIcon",
  viewBox: "0 0 12 12",
  path: (
    <path
      fill="none"
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="1.3"
      d="M3 3l6 5.943M3 9l5.94-6"
    />
  ),
});

describe("createIcon", () => {
  expect.assertions(1);

  it("should render", () => {
    const { asFragment } = render(<Test />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should not have a11y violations", async () => {
    await testA11y(<Test />);
  });
});
