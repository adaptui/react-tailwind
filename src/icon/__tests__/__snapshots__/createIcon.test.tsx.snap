// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createIcon should render 1`] = `
<DocumentFragment>
  <svg
    class="lib:w-em lib:h-em lib:inline-block lib:leading-em lib:flex-shrink-0 lib:text-current lib:align-middle"
    data-testid="testid-icon"
    focusable="false"
    viewBox="0 0 12 12"
  >
    <path
      d="M3 3l6 5.943M3 9l5.94-6"
      fill="none"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="1.3"
    />
  </svg>
</DocumentFragment>
`;
